lark/parser_frontends.py

```diff
@@ -27,57 +26,108 @@ def lex(self, lexer_state, parser_state):
+        parser = LALR_Parser.deserialize(data['parser'], memo, callbacks, options.debug)
+        parser_conf.callbacks = callbacks
+
+        terminals = [item for item in memo.values() if isinstance(item, TerminalDef)]
+
+        lexer_conf.callbacks = _get_lexer_callbacks(options.transformer, terminals)
+        lexer_conf.re_module = regex if options.regex else re
+        lexer_conf.use_bytes = options.use_bytes
+        lexer_conf.g_regex_flags = options.g_regex_flags
+        lexer_conf.skip_validation = True
+        lexer_conf.postlex = options.postlex
+
+        return ParsingFrontend(lexer_conf, parser_conf, options, parser=parser)
+
+
+
+
+class ParsingFrontend(Serialize):
+    __serialize_fields__ = 'lexer_conf', 'parser_conf', 'parser', 'options'
+
+    def __init__(self, lexer_conf, parser_conf, options, parser=None):
+        self.parser_conf = parser_conf
+        self.lexer_conf = lexer_conf
+        self.options = options
+
+        # Set-up parser
+        if parser:  # From cache
+            self.parser = parser
         else:
-            raise ConfigurationError('Unknown lexer: %s' % lexer)
-    elif parser == 'cyk':
-        if lexer == 'standard':
-            return CYK
+            create_parser = {
+                'lalr': create_lalr_parser,
+                'earley': create_earley_parser,
+                'cyk': CYK_FrontEnd,
+            }[parser_conf.name]
+            self.parser = create_parser(lexer_conf, parser_conf, options)
+
+        # Set-up lexer
+        self.skip_lexer = False
+        if lexer_conf.name in ('dynamic', 'dynamic_complete'):
+            self.skip_lexer = True
+            return
+
+        try:
+            create_lexer = {
+                'standard': create_traditional_lexer,
+                'contextual': create_contextual_lexer,
+            }[lexer_conf.name]
+        except KeyError:
+            assert issubclass(lexer_conf.name, Lexer), lexer_conf.name
+            self.lexer = _wrap_lexer(lexer_conf.name)(lexer_conf)
         else:
-            raise ConfigurationError('CYK parser requires using standard parser.')
-    else:
-        raise ConfigurationError('Unknown parser: %s' % parser)
+            self.lexer = create_lexer(lexer_conf, self.parser, lexer_conf.postlex)
+
+        if lexer_conf.postlex:
+            self.lexer = PostLexConnector(self.lexer, lexer_conf.postlex)
 
 
-class _ParserFrontend(Serialize):
     def _parse(self, start, input, *args):
         if start is None:
-            start = self.start
+            start = self.parser_conf.start
             if len(start) > 1:
                 raise ConfigurationError("Lark initialized with more than 1 possible start rule. Must specify which start rule to parse", start)
             start ,= start
         return self.parser.parse(input, start, *args)
 
+    def parse(self, text, start=None):
+        if self.skip_lexer:
+            return self._parse(start, text)
+
+        lexer = LexerThread(self.lexer, text)
```

Comment:
Here. Also, other files outside of this commit also had this problem.

Reply:
Hmm okay.. I mean, you see the init right there. But I guess it can't hurt to be more explicit.

Reply:
This is in general a 'problem' in this library: many things that could be justified into the same variable name:

`parser`:

- `Lark`
- `ParsingFrontend`
- `LALR_Parser`/`Eearley`/`XEealey`/`CYK` instances
- `LALR_Parser`/`Eearley`/`XEealey`/`CYK` classes
- the strings specifying  a parser

`lexer`

- `Lexer` class
- `Lexer` instance
- `LexerState` instance (but I don't think this occurs anywhere)
- `LexerThread`
- the strings specifying a lexer

`options`/`config`

- `LarkOptions` instance
- `LexerConfig`
- `ParserConfig`

while it is almost always clear from context which of these is meant, it doesn't hurt to be specific.

Reply:
Btw, I am not saying 'rename all variables  that might be confusing'. Just be aware of especially problematic sections.

as example, it should clear if we are looking at an implementation for `parser.parse`, `parser.parser.parse` or `parser.parser.parser.parse` if there is a difference in what we are supposed to give them (e.g. `text` vs `lexer_thread`).